# Tarea 2 Parte 2

In [1]:
import numpy as np

## Ejercicio 1

En ML usamos la entropía cruzada como una forma de medir que tan bueno es un modelo de variables discretas a través de comparar la distribución de probabilidad que el modelo produce o predice, vs la distribución de probabilidad real dada por los datos de entrenamiento.

Podemos ver el siguiente ejemplo que define la forma en que se calcula la entropía cruzada y nos muestra un caso específico:

<img src="https://image.slidesharecdn.com/publishintroductiontodeeplearninginpythonandmatlab1-160502102437/95/introduction-to-deep-learning-in-python-and-matlab-54-638.jpg?cb=1462185644">

En este caso interpretamos así: El problema consiste en un modelo o algoritmo de ML que debe producir un vector de 3 elementos indicando la probabilidad de que ciertos datos X pertenezcan a una de 3 categorías.

* El modelo de ML produce un vector que indica que estima un 70% de probabilidad de que se trate de la categoría 0, 20% de probabilidad de que se trate de la categoría 1 y 10% de que se trate de la categoría 2.
* Los datos reales nos dicen que se trataba de un caso donde con total certeza se sabe que se trata de la categoría 0
* La entropía cruzada(a calcular en el ejercicio) nos indica que tan buena es la estimación del modelo, una EC de 0 es un modelo perfecto(en este caso un modelo que predice 100% de prob para la clase 0)

**Nota** 
* Aun que para calcular la entropía usamos logaritmos en base 2, en ML para calcular la entropía cruzada se usa logaritmo natural ya que con este se cumple el proposito **estimar que tanto se alejan las predicciones del modelo de ML de los datos reales** y es comun]mente mas rápido de calcular en la computadora.
* Ya que estamos trabajando con vectores que representan distribuciones de probabilidad , podemos toparnos con lo que se conoce como : **sparse vectors**(vectores donde la mayoría de elementos son 0), esto puede producir problemas ya que le logaritmo de 0 no esta definido, tu solución debe tomar en cuenta esto y evitar que devuelva "nan"

In [2]:
def cross_entropy(Y,Y_hat):
    ## tu codigo aqui (~ 1 lineas de codigo)
    #return -1*(Y*np.log(Y_hat)) - ((1 - Y)*np.log(1 - Y_hat))
    return -1 * (np.sum(Y[Y_hat>0]*np.log(Y_hat[Y_hat>0])))
    ##

y  = np.array([1.0,0,0])
y_hat = np.array([0.7,0.2,0.1])

cross_entropy(y,y_hat)

0.35667494393873245


# Magnitud de un Vector(norma o módulo)
 

Matemáticamente la magnitud de un vector(también conocida como norma vectorial)  nos indica  el tamaño de este , y nos sirve para tener una noción de la distancia desde un punto de referencia(origen) hasta el punto representado por el vector.


## Ejercicio 2
Crear una función que reciba como parámetro un vector x y calcule su magnitud o norma(euclidiana o L2) ,luego usarla para evaluar 2 vectores que representan los errores generados por 2 modelos de machine learning y concluir cual de los 2 modelos es mejor.

In [3]:
def magnitud(x):
    ## tu codigo aqui (~ 1 linea de codigo)
    return np.sqrt(np.sum(np.power(x,2)))
    ##
    

errores_modelo1 = np.array([1,2,1,2])
errores_modelo2 = np.array([0,1,1,2])

print(magnitud(errores_modelo1))
print(magnitud(errores_modelo2))

3.1622776601683795
2.449489742783178


## Ejercicio 3
Usando la función del ejercicio anterior, crea otra función normalizar(x) que reciba de parámetro un vector x aplique normalización sobre este, el resultado debe ser un nuevo vector del tamaño de x cuya magnitud es igual a 1.0

In [4]:
def normalizar(x):
    ## tu codigo aqui (~ 3 linea de codigo)
    v = np.divide(x,magnitud(x))
    return v
    ##
    

print(magnitud(normalizar(errores_modelo1)))
print(normalizar(errores_modelo2))

1.0
[0.         0.40824829 0.40824829 0.81649658]


# Producto Punto(escalar,interno,interior) y Ortogonalidad



## Ejercicio 4
### Aplicado en DS

Se tiene una red neuronal sencilla(y simplificada) como la de la siguiente imagen:
<img src="https://www.oreilly.com/library/view/practical-convolutional-neural/9781788392303/assets/246151fb-7893-448d-b9bb-7a87b387a24b.png">

Donde:
* INPUT LAYER: un vector X de tamaño = 2 que representa los datos de entrada
* HIDDEN_LAYER :capa oculta con 2 neuronas definidas por los vectores:
    * HL1 = [0.25,0.37]
    * HL2 = [-8,14]
* OUTPUT_LAYER = capa de salida definida por el vector [4,9]

Crear una funcion `neural_network(X)` para calcular:
* Calcule la salida de cada neurona en la capa intermedia aplicada a la capa de entrada.
* Use el resultado del paso anterior como entrada para la neurona en la capa de salida

Asumiendo que cada neurona identifica la similitud entre su entrada y la caracteística que representa concluir:
* Para cada vector de entrada Xi , cual neurona intermedia busca la característica que mas se parece a X.
* Cual vector de entrada Xi produce una activación alta(salida alta) en la capa de salida.

In [5]:
# Aplicar la red neuronal sobre los siguientes datos X

X1 = np.array([0.50,0.72])
X2 = np.array([-4,7])
X3 = np.zeros_like(X2)
x4 = np.ones_like(X1)
X5 = np.random.randn(X1.shape[0])

In [14]:
## tu codigo aqui:
def neural_network(x):
    OL = [4,9]
    HL1 = [0.25,0.37]
    HL2 = [-8,14]
    
    S1 = np.dot(x,HL1)
    S2 = np.dot(x,HL2)
    S = np.array([S1,S2])
    return S, np.dot(OL,S)

print('X1: ', neural_network(X1))
print('X2: ', neural_network(X2))
print('X3: ', neural_network(X3))
print('x4: ', neural_network(x4))
print('X5: ', neural_network(X5))

X1:  (array([0.3914, 6.08  ]), 56.2856)
X2:  (array([  1.59, 130.  ]), 1176.36)
X3:  (array([0., 0.]), 0.0)
x4:  (array([0.62, 6.  ]), 56.48)
X5:  (array([ 0.3054224 , -4.44539992]), -38.78690967195477)


# Conclusión

**Para cada vector de entrada Xi , ¿cual neurona intermedia busca la característica que mas se parece a X?**

X1: Para este vector, la neurona intermedia HL2 es la que busca una característica más parecida a X1 con un valor de 6.08.

X2: Para este vector, la neurona intermedia HL2 es la que busca una característica más parecida a X2 con un valor de 130.

X3: Para este vector, las dos neuronas intermedia HL1 y HL2 son iguales ante X1 ya que su producto punto es de 0.

x4: Para este vector, la neurona intermedia HL2 es la que busca una característica más parecida a X4 con un valor de 6.

X5: Para este vector, la neurona intermedia HL1 es la que busca una característica más parecida a X5 con un valor de 0.3054224.

**¿Cual vector de entrada Xi produce una activación alta(salida alta) en la capa de salida?**

El vector X2 es el que da una mayor activación con un valor de 1176.36

## Ejercicio 5 (aplicado en DS)
La correlación cruzada es una medida de similitud entre 2 funciones como resultado de "desplazar" una sobre la otra, comunmente es usada para encontrar características relevantes en una función desconocida o no controlada.

En procesamiento de señales por ejemplo es usada para buscar y/o filtrar en una señal que varia en el tiempo cierta caracaterística de interés. Aveces es llamada también "sliding dot product" consiste en aplicar en cada punto de una función F el producto punto con cierta función G(comunmente mas corta) y luego "deslizar" G a un nuevo punto de la función, el resultado es una nueva función H que se interpreta como :**cuanto se parece** en cada punto la función F a la característica G. 

<img src="https://i.makeagif.com/media/11-25-2015/LZ9Ufj.gif"  height="250" width="300">

En este ejercicio usamos correlación cruzada para calcular las medias moviles promediando 3 puntos que ya vimos en otro ejemplo:

El primer paso es definir la función G que define el "filtro" a aplicar, para este caso consiste simplemente en un vector con 3 elementos donde cada elemento corresponde a 1/3, luego debemos aplicar el producto punto sobre cada punto de la función o datos originales(en este ejemplo llamados x) a traves aplicar el producto punto en cada elemento "corriendolo" de uno en uno

In [56]:
ruido = 0.1*np.random.randn(15) #el ruido comunmente se debe a aleatoriedad o captura no exacta de info.
x = np.linspace(0,2*np.pi,15) + ruido
x = np.sin(x)

# Version original usando FOR y promedios
xs =  x.copy()

for n in range(1,len(x)-1):
    xs[n] = (x[n-1] + x[n] + x[n+1])/3
    
print(xs)

# Version final usando producto punto
filtro = np.array([1/3,1/3,1/3])

## tu codigo aqui (~ 3 linea de codigo)
x1=np.arange(0,len(x) -2) #arreglo de indices 1 
x2=np.arange(1,len(x) -1) #arreglo de indices 2
x3=np.arange(2,len(x)) #arreglo de indices 3

#combinamos los indices para crear una matriz en donde cada fila posee una columna de la forma x1, x2, x3
xs=np.c_[x[x1],x[x2],x[x3]]

#sacamos el producto punto entre las filas de la matriz y el vector de filtro
xs= np.dot(xs,filtro)

#agregamos al inicio el elemento x[0]
xs= np.append(x[0],xs)

#agregamos al final el elemento x[-1]
xs= np.append(xs,x[-1])

#imprimimos resultado
print(xs)




[ 0.04863352  0.40935378  0.72109052  0.91278238  0.91342817  0.75041212
  0.44162847 -0.00434153 -0.41145496 -0.76343997 -0.88753208 -0.90298082
 -0.73899538 -0.43390492 -0.03457344]
[ 0.04863352  0.40935378  0.72109052  0.91278238  0.91342817  0.75041212
  0.44162847 -0.00434153 -0.41145496 -0.76343997 -0.88753208 -0.90298082
 -0.73899538 -0.43390492 -0.03457344]


# Operadores logicos en vectores

## Ejercicio 6
Dado el vector x, usar operadores lógicos sobre vectores y acceso a  elementos usando vectores booleanos , escribir un programa que calcule eun nuevo vector z conteniendo el valor absoluto de el vector x.

**Nota** No se puede usar np.abs() ni ciclos

**Tip** usar un vector booleano para saber que elementos son negativos y deben ser multiplicados por -1

In [2]:
x = np.linspace(-2,2,9)


## tu codigo aqui (~ 4 linea de codigo)
print(x)
z = np.copy(x)
z[z<0]*=-1
print(z)

[-2.  -1.5 -1.  -0.5  0.   0.5  1.   1.5  2. ]
[2.  1.5 1.  0.5 0.  0.5 1.  1.5 2. ]




## Ejercicio 7
Implementar la función:
$$h(x) = \begin{cases}0 & x< 0\\1 & 0<=x

$$h(x) = \begin{cases}0 & x< 0\\1 & 0<=x <=1\\0 & x> 1\end{cases}$$

Esta debe funcionar para vectores de cualquier tamaño x:

```
def h(x):
    
    ...
    
```
**nota** debe ser implementada sin ciclos o ifs 

In [4]:
def h(x):
    ## tu codigo aqui (~ 3 linea de codigo)
    x[(x>=0) & (x<=1)] = 1
    x[x>1] = 0
    x[x<0] = 0
    h = x
    ##
    return h

x = np.array([0.1,-2,0.5,5])
print(h(x))

[1. 0. 1. 0.]


Para casos donde se requiere comportamiento parecido a este pero con mas condiciones, NumPy provee la función **np.select** , la descripción de esta función casi siempre es mas complicada de lo que debería y dificulta entenderla, vamos a buscar entenderla bajo un ejemplo.

Básicamente select se basa en :
* una lista de condiciones A
* una lista de valores a tomar según estas condiciones B
* un valor de  resultado default cuando ninguna de las condiciones en la  lista se cumple.
Estas 2 listas deben ser del mismo tamañaño.

# Select y performance

Usando select implementar una función k(x) con la siguiente definición:
$$k(x) = \begin{cases}-x & x< 0\\x^{3}  &0<=x<1\\x^{2}  &1<=x<2\\4  &otherwise\end{cases}$$



Aun que select es muy conveniente y útil , no es la opción mas eficiente esto debido a que evalua todas las condiciones y todos los resultados, en programación en general buscamos escribir los programas de la manera mas eficiente posible y evitar calculos innecesarios, esto se vuelve vital en ciencia de datos cuando procesamos grandes volúmenes de información, por eso NumPy nos provee la alternativa a select llamada **piecewise** que funciona de manera similar .

Este funciona similar a select pero en vez de calcular todos los posibles resultados, calcula solo aquellos para los que la condición es True e ignora los False,sintácticamente piecewise requiere que los "resultados" sesan calcuados usando una lista de funciones por lo cual todos los resultados deben estar contenidos en una función  y el objeto función ser enviado a piecewise (si la función solo se utiliza una vez para este propósito, se puede usar funciones anónimas o lambda).

## Ejercicio 8

Investigar piecewise:
*  Usarlo para implementar la función anterior de manera eficiente
*  Usar piscewise para implementar la función:

$$m(x) = \begin{cases}e^{2x} & x< 0\\1  &0<=x<1\\e^{1-x}  &x>=1\\\end{cases}$$

Nuevamente, sin utilizar ciclos ni ifs

In [7]:
#implementamos la funcion k utilizando select
def k(x):
    arrCondiciones = [x<0, (x>=0)&(x<1), (x >=1) & (x<2)]
    arrResultados = [x*-1, np.power(x,3), np.power(x,2)]
    return np.select(arrCondiciones, arrResultados, default = 4)

#invocamos la funcion
x = np.array([-1,0,0.5,1,1.5,2,3])
print(k(x))

# Implementamos la funcion anterior usando piecewise
def k(x):
    arrCondiciones = [x<0, (x>=0)&(x<1), (x >=1) & (x<2), x>=2]
    arrFunciones = [lambda x: x*-1, lambda x: np.power(x,3), lambda x: np.power(x,2), 4]
    return np.piecewise(x,arrCondiciones, arrFunciones)

#invocamos la funcion
x = np.array([-1,0,0.5,1,1.5,2,3])
print(k(x))

# implementamos la funcion m usando piecewise
def m(x):
    arrCondiciones = [x<0, (x>=0)&(x<1), x>=1]
    arrFunciones = [lambda x: np.exp(2*x), 1, lambda x: np.exp(1-x)]
    return np.piecewise(x,arrCondiciones,arrFunciones)

#invocamos la funcion
x = np.array([-1,0,0.5,1,1.5,2,3])
print(m(x))

[1.    0.    0.125 1.    2.25  4.    4.   ]
[1.    0.    0.125 1.    2.25  4.    4.   ]
[0.13533528 1.         1.         1.         0.60653066 0.36787944
 0.13533528]
